controllers/parser.go

```diff
@@ -0,0 +1,357 @@
+				}
+				routeRoute, err := generateRoute(
+					clusterName,
+					corsPolicy,
+					&generateRouteOpts{
+						RewriteRegex: &rewriteOpts,
+						QoS:          methodOpts.QoS,
+					},
+				)
+				if err != nil {
+					return err
+				}
+
+				rt.Action = routeRoute
+			}
+
+			for _, vh := range envoyConfiguration.GetVirtualHosts() {
+				if err := envoyConfiguration.AddRouteToVHost(vh.Name, rt); err != nil {
+					return fmt.Errorf("failure adding redirect route: %w", err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func generateRouteMatch(path string, method string, pathParameters map[string]types.ParamSchema, corsPolicy *route.CorsPolicy) *route.RouteMatch {
+	headerMatcherConfig := []*route.HeaderMatcher{
+		types.GetHeaderMatcherConfig([]string{strings.ToUpper(method)}, corsPolicy != nil),
+	}
+
+	routeMatcherBuilder := types.NewRouteMatcherBuilder(path, pathParameters)
+	return routeMatcherBuilder.GetRouteMatcher(headerMatcherConfig)
+}
+
+func generateRedirect(redirectOpts *options.RedirectOptions) (*route.Route_Redirect, error) {
+	if redirectOpts == nil {
+		return nil, nil
+	}
+
+	redirect, err := types.NewRouteRedirectBuilder().
+		HostRedirect(redirectOpts.HostRedirect).
+		PortRedirect(redirectOpts.PortRedirect).
+		SchemeRedirect(redirectOpts.SchemeRedirect).
+		RegexRedirect(redirectOpts.RewriteRegex.Pattern, redirectOpts.RewriteRegex.Substitution).
+		PathRedirect(redirectOpts.PathRedirect).
+		ResponseCode(redirectOpts.ResponseCode).
+		StripQuery(redirectOpts.StripQuery).
+		ValidateAndReturn()
+
+	if err != nil {
+		return nil, err
+	}
+
+	return redirect, nil
+}
+
+func generateCORSPolicy(corsOpts *options.CORSOptions) (*route.CorsPolicy, error) {
+	if corsOpts == nil {
+		return nil, nil
+	}
+
+	return types.GenerateCORSPolicy(
+		corsOpts.Origins,
+		corsOpts.Methods,
+		corsOpts.Headers,
+		corsOpts.ExposeHeaders,
+		corsOpts.MaxAge,
+		corsOpts.Credentials,
+	)
+}
+
+func getUpstreamHost(upstreamOpts *options.UpstreamOptions) (hostname string, port uint32) {
+	if upstreamOpts.Service != nil {
+		return fmt.Sprintf("%s.%s.svc.cluster.local.", upstreamOpts.Service.Name, upstreamOpts.Service.Namespace), upstreamOpts.Service.Port
+	}
+	return upstreamOpts.Host.Hostname, upstreamOpts.Host.Port
+}
+
+// each cluster can be uniquely identified by dns name + port (i.e. canonical Host, which is hostname:port)
+func generateClusterName(name string, port uint32) string {
+	return fmt.Sprintf("%s-%d", name, port)
+}
+
+// Can be moved to operationID, but generally we just need unique string
+func generateRouteName(path string, method string) string {
+	return fmt.Sprintf("%s-%s", path, strings.ToUpper(method))
+}
+
+func generateRoutePath(prefix, path string) string {
+	if prefix == "" {
+		return path
+	}
+
+	// Avoids path joins (removes // in e.g. /path//subpath, or //subpath)
+	return fmt.Sprintf(`%s/%s`, strings.TrimSuffix(prefix, "/"), strings.TrimPrefix(path, "/"))
+}
+
+type generateRouteOpts struct {
```

Comment:
These are only 2 options to pass to the functions, I don't think that we need this struct for this.